docs/non_native_field_arithmetic.md

```diff
@@ -0,0 +1,164 @@
+a_0\rbrack$ with
+$a_i \in \lbrack 0, 2^B)$ and $a = \sum\limits_{0 \leq i < 4}{a_i \cdot 2^{iB}}$ . Given $a,b,q,p' \in \lbrack 0, 2^T)$, we can rewrite
+the equation
+$a \cdot b + p' \cdot q$ as follow:
+
+$$
+\begin{aligned}
+a \cdot b + p' \cdot q & = \lbrack a_3, a_2, a_1, a_0\rbrack \cdot \lbrack b_3, b_2, b_1, b_0\rbrack + \lbrack p'_ 3, p'_ 2, p'_ 1, p'_
+0\rbrack \cdot \lbrack q_3, q_2, q_1, q_0\rbrack \\
+& =(a_{0}\cdot b_{0}+q_{0}\cdot p_{0}^{\prime})\cdot2^{0B}+ \\
+& \quad \ (a_{1}\cdot b_{0}+a_{0}\cdot b_{1}+q_{1}\cdot p_{0}^{\prime}+q_{0}\cdot p_{1}^{\prime})\cdot2^{1B}+ \\
+& \quad \ (a_{2}\cdot b_{0}+a_{0}\cdot b_{2}+a_{1}\cdot b_{1}+q_{2}\cdot p_{0}^{\prime}+q_{0}\cdot p_{2}^{\prime}+q_{1}\cdot p_
+{1}^{\prime})\cdot2^{2B}+ \\
+& \quad \ (a_{3}\cdot b_{0}+a_{0}\cdot b_{3}+a_{1}\cdot b_{2}+a_{2}\cdot b_{1}+q_{3}\cdot p_{0}^{\prime}+q_{0}\cdot p_{3}^{\prime}+q_
+{1}\cdot p_{2}^{\prime}+q_{2}\cdot p_{1}^{\prime})\cdot2^{3B} \\
+& =t_{0}\cdot2^{0B}+t_{1}\cdot2^{1B}+t_{2}\cdot2^{2B}+t_{3}\cdot2^{3B}
+\end{aligned}
+$$
+
+We can ignore $a_i \cdot b_j$ with $i+j \geq 4$ because it is divisible by $2^T = 2^{4B}$. Now we have:
+
+$$
+\begin{array}{l}
+{{t_{0}\in\lbrack 0,2^{2B+1})\mathrm{}}}\\
+{{t_{1}\in\lbrack 0,2^{2B+2})\mathrm{}}}\\
+{{t_{2}\in\lbrack 0,2^{2B+3})\mathrm{}}}\\
+{{t_{3}\in\lbrack 0,2^{2B+3})\mathrm{}}}\end{array}
+$$
+
+Let $t = t_{0}\cdot2^{0B}+t_{1}\cdot2^{1B}+t_{2}\cdot2^{2B}+t_{3}\cdot2^{3B}$. Note that $t$ itself does not need to be less than
+$2^T$. Thus, if $a \cdot b + p' \cdot q = r \mod 2^T$, it is not guaranteed that $t = r$. However, the last $T$ bits of $t$ are
+certainly going to equal $r$ (the reason why we choose $2^T$ as our modulus).
+
+![non_native_field_arithmetic](attachments/non_native_field_arithmetic.png)
+
+With $r = \lbrack r_3, r_2, r_1, r_0\rbrack$, we have $r_0$ is the last $B$ bits of $y_0 = t_0 + 0$, i.e $y_0-r_0 = z_0 \cdot 2^B$ for
+some
+non-negative $z_0$. Similarly, $r_1$ is the $B$ bits of $y_1 = t_1 + z_0$ with $y_1 - r_1 = z_1 \cdot 2^B$. We can write the all the
+limbs of $r$ in a similar fashion:
+
+$$
+\begin{array}{l l}
+{{y_{0}=t_{0}+0,}}&{{y_{0}-r_{0}=z_{0}\cdot2^{B}}}\\ 
+{{y_{1}=t_{1}+z_{0},}}&{{y_{1}-r_{1}=z_{1}\cdot2^{B}}}\\ 
+{{y_{2}=t_{2}+z_{1},}}&{{y_{2}-r_{2}=z_{2}\cdot2^{B}}}\\ 
+{{y_{2}=t_{3}+z_{2},}}&{{y_{3}-r_{3}=z_{3}\cdot2^{B}}}
+\end{array}
+$$
+
+Removing the intermediate variables $y_i$, we get:
+
+$$
+\begin{array}{l}
+{{t_{0}+0-r_{0}=z_{0}\cdot2^{B}}}\\ 
+{{t_{1}+z_{0}-r_{1}=z_{1}\cdot2^{B}}}\\ 
+{{t_{2}+z_{1}-r_{2}=z_{2}\cdot2^{B}}}\\
+{{t_{3}+z_{2}-r_{3}=z_{3}\cdot2^{B}}}
+\end{array}
+$$
+
+To summarize, for a given $(\lbrack a_3,a_2,a_1,a_0\rbrack, \lbrack b_3, b_2, b_1, b_0\rbrack, \lbrack p'_3, p'_2, p'_1, p'_0\rbrack,
+\lbrack q_3, q_2, q_1, q_0\rbrack, \lbrack r_3,r_2,r_1,r_0\rbrack,
+t_3,t_2,t_1,t_0, z_3,z_2,z_1,z_0)$ that satisfies the 4 constraints above, along with constraints that ensures $p'=2^T-p$ (note that
+$p$ is an odd prime),
+
```

Comment:
need linking word here

Reply:
I don't understand this, can you add a suggestion ?

Reply:
add "and"